### ISSUE TITLE
runtime: failure to getPageSize on netbsd-386

### ISSUE THREAD

**Issue by bradfitz** [golang/go#22914]
I'm trying to revive the NetBSD port.

We have passing NetBSD 8-BETA builders for amd64 now.

The NetBSD i386 kernel fails to get networking on GCE, but the 64-bit NetBSD kernel can run 32-bit binaries (like FreeBSD and unlike OpenBSD), and that's how we used to run the netbsd-386 builder, so I'm trying that again.

But 32-bit binaries fail to get the page size on a 64-bit kernel:

```
Running /tmp/workdir/go/src/all.bash with args ["/tmp/workdir/go/src/all.bash"] and env ["PWD=/" "HOME=/" "PATH=/usr/pkg/bin:/usr/pkg/sbin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/pkg/sbin:/usr/pkg/bin:/usr/X11R7/bin:/usr/X11R6/bin:/usr/local/sbin:/usr/local/bin" "_rc_original_stdout_fd=7" "GOROOT_BOOTSTRAP=/tmp/workdir/go1.4" "_rc_original_stderr_fd=8" "RC_PID=  10" "_rc_pid=2" "_rc_postprocessor_fd=9" "WORKDIR=/tmp/workdir" "GO_BUILDER_NAME=netbsd-386-8branch" "GOARCH=386" "GOHOSTARCH=386"] in dir /tmp/workdir/go/src
Building Go toolchain1 using /tmp/workdir/go1.4.
Building Go bootstrap cmd/go (go_bootstrap) using Go toolchain1.
Building Go toolchain2 using go_bootstrap and Go toolchain1.
fatal error: failed to get system page size
runtime: panic before malloc heap initialized

runtime stack:     
runtime.throw(0x82bf412, 0x1e)
        /tmp/workdir/go/src/runtime/panic.go:616 +0x65 fp=0xffffe718 sp=0xffffe70c pc=0x806e025
runtime.mallocinit()        
        /tmp/workdir/go/src/runtime/malloc.go:232 +0x3ba fp=0xffffe758 sp=0xffffe718 pc=0x8055c6a
runtime.schedinit()         
        /tmp/workdir/go/src/runtime/proc.go:485 +0x3c fp=0xffffe780 sp=0xffffe758 pc=0x807029c
runtime.rt0_go(0xffffe814, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /tmp/workdir/go/src/runtime/asm_386.s:297 +0x224 fp=0xffffe784 sp=0xffffe780 pc=0x8091ef4
go tool dist: FAILED: /tmp/workdir/go/pkg/tool/netbsd_386/go_bootstrap install -gcflags=all= -ldflags=all= -i cmd/asm cmd/cgo cmd/compile cmd/link: exit status 2
```

Obviously not correct, but falling back to 4KB instead of returning 0 on sysctl failure seems to make it get further:

```patch
bradfitz@gdev:~/go/src$ git di
diff --git a/src/runtime/os_netbsd.go b/src/runtime/os_netbsd.go
index 39e91ee..a35bd1c 100644
--- a/src/runtime/os_netbsd.go
+++ b/src/runtime/os_netbsd.go
@@ -106,6 +106,9 @@ func getPageSize() uintptr {
        if ret >= 0 {
                return uintptr(out)
        }       
+       if GOOS == "netbsd" {
+               return 4 << 10  
+       }               
        return 0        
 }              

```

Now instead of crashing, `make.bash` seems to hang forever, consuming CPU forever in `/tmp/workdir/go/pkg/tool/netbsd_386/go_bootstrap install`.

Ideas welcome.

/cc @bsiegert @aclements @krytarowski @ianlancetaylor

**Comment by bradfitz** [comment]
If we don't get netbsd-386 working again, we'll just document it in the release notes. That is #22911. At least having any NetBSD working again would be an improvement. (Go 1.9 didn't run on NetBSD at all.)

**Comment by bradfitz** [comment]
Debugging this would be easier once ssh is enabled for NetBSD. That is #22872.

**Comment by krytarowski** [comment]
Panic... this looks badly. Trying to reproduce locally.

**Comment by coypoop** [comment]
hi, if I run netbsd_386/go 'ktruss -i go'
 10875      1 go       netbsd32___sysctl(0xffffffffffffe8cc, 0x2, 0xffffffffffffe8c4, 0xffffffffffffe8c8, 0, 0) = 0

I think it's not failing, and strongly suspect that sysctl in sys_netbsd_386.s is wrong, but I'm not very knowledgeable in x86 asm

**Comment by coypoop** [comment]
copying the freebsd/386 sysctl implementation as-is fixes this

**Comment by bradfitz** [comment]
@coypoop, thanks!

For the record, the diff is:

```patch
localhost# git diff
diff --git a/src/runtime/sys_netbsd_386.s b/src/runtime/sys_netbsd_386.s
index af8c3aa485..8ee7b96ce7 100644
--- a/src/runtime/sys_netbsd_386.s
+++ b/src/runtime/sys_netbsd_386.s
@@ -375,10 +375,12 @@ TEXT runtime·sysctl(SB),NOSPLIT,$28
        MOVSL                           // arg 6 - newlen
        MOVL    $202, AX                // sys___sysctl
        INT     $0x80
-       JCC     3(PC)
+       JAE     4(PC)
        NEGL    AX
+       MOVL    AX, ret+24(FP)
        RET
        MOVL    $0, AX
+       MOVL    AX, ret+24(FP)
        RET
 
 GLOBL runtime·tlsoffset(SB),NOPTR,$4
```

@rsc fixed that for freebsd/386 in 25f6b02ab0db8e22994a3a68f4cb2857b788a63f where he wrote:

>    If this breaks another system, the bug is almost certainly in the
>    sys_$GOOS_$GOARCH.s file, since the rest of the CL is tested
>    by the combination of the above systems.

Indeed.
